Write StatusSummaries and metrics at startup

mrtracy opened the pull request:
The server will now aggressively record node status summaries and metrics
immediately at startup; this alleviates the current poor experience in the UI,
where the server reports having zero nodes for the first ten seconds (a similar
delay occurs when other nodes are added, but are not immediately present in the
UI.)

Status summaries are also recorded as soon as stores are bootstrapped on new
nodes; this prevents nodes from existing with zero stores for a lengthy period.

Finally, this bug repairs a UI issue where a node is returned which has zero
stores (possible before stores have finished bootstrapping). This caused a
null-reference in the UI which blanked out the "Nodes" page. We discussed
several methods of preventing this particular null reference, including
modifications to the way protobuffers are marshalled to JSON; in the end, we
agreed that handling the possible null on the client side was the best option
for now. I have logged issue #4849 to follow up on this in the future.

Fixes #4747

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4850)

<!-- Reviewable:end -->

Commit:
Write StatusSummaries and metrics at startup

The server will now aggressively record node status summaries and metrics
immediately at startup; this alleviates the current poor experience in the UI,
where the server reports having zero nodes for the first ten seconds (a similar
delay occurs when other nodes are added, but are not immediately present in the
UI.)

Status summaries are also recorded as soon as stores are bootstrapped on new
nodes; this prevents nodes from existing with zero stores for a lengthy period.

Finally, this bug repairs a UI issue where a node is returned which has zero
stores (possible before stores have finished bootstrapping). This caused a
null-reference in the UI which blanked out the "Nodes" page. We discussed
several methods of preventing this particular null reference, including
modifications to the way protobuffers are marshalled to JSON; in the end, we
agreed that handling the possible null on the client side was the best option
for now. I have logged issue #4849 to follow up on this in the future.

Fixes #4747

petermattis commented:
LGTM

---

Review status: 0 of 9 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[server/node.go, line 576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzQkPsWTKDedewG7c2:-KBzQkPsWTKDedewG7c3:628220361) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L576)):*
Was there a reason to move this code from `Server` to `Node`?

---

*[server/node.go, line 611 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzPmgDs8vshWyCziMT:-KBzPmgDs8vshWyCziMU:977085152) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L611)):*
Is there a reason you're logging json here vs using `%+v`? 

---

*[server/server.go, line 315 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzPy4NoVmM4Pf10P4k:-KBzPy4NoVmM4Pf10P4l:-450060176) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/server.go#L315)):*
`s/write/Write/g`

---

*[server/server.go, line 317 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzQy8hTr_j4wnbP4kV:-KBzQy8hTr_j4wnbP4kW:-2037356971) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/server.go#L317)):*
I would have done this call in `startWriteSummaries`. That is, before you start looping in `startWriteSummaries`, perform an initial call to `writeSummaries`. This would ensure that you never have multiple calls to `writeSummaries` happening at once which seems modestly cleaner.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-:-KBzRFLDXgAEYURgdbLr:676834233)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[server/node.go, line 622 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzTd0KSAYZt1D_Vppl:-KBzTd0KSAYZt1D_Vppm:-86682823) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L622)):*
why marshal this to json at all here?

---

*[ui/ts/models/status.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzStAQgL3hTLBWNXxK:-KBzStARWgwqVWuMKHXb:981038730) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/ui/ts/models/status.ts#L69)):*
looks like a fair amount of duplication happening here. perhaps just setting the key to an empty array if it is unset would be cleaner.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 4 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[server/node.go, line 576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzQkPsWTKDedewG7c2:-KBzcpd4-0Y74Gv2byc8:1406041450) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L576)):*
Yes, I wanted to write status summaries at the end of `bootstrapStores`, which is entirely inside of Node and doesn't have access to the server.  

After moving it, I also think it makes more sense inside of Node anyway.

---

*[server/node.go, line 611 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzPmgDs8vshWyCziMT:-KBzd9uwnMmpbjaDSpP3:1979475524) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L611)):*
This has been around for a long time, but I'm assuming the JSON output just visually looks better than the %+v output.

---

*[server/node.go, line 622 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzTd0KSAYZt1D_Vppl:-KBzdMwp-9UJMe-_RwNG:-1287300556) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/node.go#L622)):*
I'm assuming it just looks better, and that's why it was done.  It's only done at log.V(1).

---

*[server/server.go, line 315 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzPy4NoVmM4Pf10P4k:-KBzdZ5mCqMfXMbGugum:-1808746638) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/server.go#L315)):*
That probably should not be `/g`.  Done.

---

*[server/server.go, line 317 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzQy8hTr_j4wnbP4kV:-KBzdjz1KnLko--r974K:1861571991) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/server/server.go#L317)):*
I originally had it that way, but I thought people might think this was cleaner.  CODE REVIEW MULTIVERSE.  I've changed it.

---

*[ui/ts/models/status.ts, line 69 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850#-KBzStAQgL3hTLBWNXxK:-KBzgpSq7e6W6Kf1lSJl:-1025158488) ([raw file](https://github.com/cockroachdb/cockroach/blob/0442ee29e213b7d611d0a31fe44a08c8fb5a9c98/ui/ts/models/status.ts#L69)):*
If you're okay with that, it is indeed significantly cleaner. I was worried that modifications during the query were conceptually worse in your opinion.

Changed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 9 files at r1, 4 of 5 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4850)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Nice work Matt. :100:
